Libraries/Cordio/wsf/sources/targets/baremetal/wsf_nvm.c

```diff
@@ -402,14 +421,76 @@ bool_t WsfNvmEraseData(uint64_t id, WsfNvmCompEvent_t compCback)
 /*************************************************************************************************/
 void WsfNvmEraseDataAll(WsfNvmCompEvent_t compCback)
 {
-  for (uint32_t eraseAddr = WSF_NVM_START_ADDR; eraseAddr < wsfNvmCb.availAddr; eraseAddr += wsfNvmCb.sectorSize)
-  {
-    PalFlashEraseSector(wsfNvmCb.availAddr, eraseAddr);
-  }
-  wsfNvmCb.availAddr = WSF_NVM_START_ADDR;
-
-  if (compCback)
-  {
-    compCback(TRUE);
-  }
+    for (uint32_t eraseAddr = WSF_NVM_START_ADDR; eraseAddr < wsfNvmCb.availAddr;
+         eraseAddr += wsfNvmCb.sectorSize) {
+        PalFlashEraseSector(wsfNvmCb.availAddr, eraseAddr);
+    }
+    wsfNvmCb.availAddr = WSF_NVM_START_ADDR;
+
+    if (compCback) {
+        compCback(TRUE);
+    }
+}
+
+/*************************************************************************************************/
+/*!
+ *  \brief  Defragment NVM
+ *  \param  copyBuf       Scratch buffer to temporarily copy NVM stored data.
+ *  \param  size          Size of copyBuf.
+ *  \return  TRUE if defragment successful. FALSE otherwise
+ * 
+ *  \note   Defragment should only be called when the storage is full and a record has been invalidated.
+ *          Furthermore, copyBuf must be at least the size of WSF NVM allocated flash.
+ */
+/*************************************************************************************************/
+bool_t WsfNvmDefragment(uint8_t *copyBuf, uint32_t size)
+{
+#if WSF_ASSERT_ENABLED == 1
+    WSF_ASSERT(copyBuf && size >= wsfNvmCb.totalSize);
+#else
+    if (!copyBuf || size < wsfNvmCb.totalSize) {
+        WSF_TRACE_INFO0("Not enough memory given to defragment NVM");
+        return FALSE;
+    }
+#endif
+
+    WsfCsEnter();
+
+
+    bool_t defragPossible = FALSE;
+    uint32_t currentOffset = WSF_NVM_START_ADDR;
+    uint32_t copyOffset = 0;
+    WsfNvmHeader_t header = { 0 };
+
+    /* Get the first header*/
+    PalFlashRead(&header, WSF_NVM_HEADER_SIZE, currentOffset);
+    while (currentOffset < wsfNvmCb.availAddr && header.id != WSF_NVM_UNUSED_FILECODE) {
+        const uint32_t fileSize = WSF_NVM_FILE_SIZE(header.len);
+
+        /*If the entry is valid copy it into the defragmentation buffer, if its reserved we can defrag it*/
+        if (header.id != WSF_NVM_RESERVED_FILECODE) {
+            PalFlashRead(&copyBuf[copyOffset], fileSize, currentOffset);
+            copyOffset += WSF_NVM_FILE_SIZE(header.len);
+        } else {
+            defragPossible = TRUE;
+        }
+
+        currentOffset += fileSize;
+        PalFlashRead(&header, WSF_NVM_HEADER_SIZE, currentOffset);
+    }
+
+    /*
+    If we defragged anything clear the sector and rewrite flash with defragmented data
+  */
+    if (defragPossible) {
+        WsfNvmEraseDataAll(NULL);
+        PalFlashWrite(copyBuf, copyOffset, wsfNvmCb.availAddr);
```

Comment:
Might want to error check and trace these last two erase/write operations as well just in case anything goes wrong here

Reply:
Both functions return void.